pennylane/ops/functions/commutator.py

```diff
@@ -101,13 +104,18 @@ def commutator(op1, op2, pauli=False):
         or insert an extra ``stop_recording()`` context (see :class:`~QueuingManager`).
 
     """
-    if pauli:
+
+    both_have_pauli_rep = op1.pauli_rep is not None and op2.pauli_rep is not None
```

Comment:
are you defining `pauli_rep` on the Pauli* classes to make this line work smoothly? If so, I'd prefer you just check their type before reaching this point

Reply:
I was thinking something more like:
```python
if pauli:
    # will do nothing if already PauliSentence
    op1 = qml.pauli.pauli_sentence(op1)
    op2 = qml.pauli.pauli_sentence(op2)
if pauli or (isinstance(op1, (PauliWord, PauliSentence)) and isinstance(op2, (PauliWord, PauliSentence))):
    # do the work in pauli-land
    return result if pauli else result.operation()
# else, do the non-pauli Operator stuff

Reply:
sorry, missed a part of the condition. updated the first `if`

Reply:
In what way is this cleaner than the current version? Am I missing something crucial? 😅

Reply:
Note also that if we dont add the trivial `.pauli_rep` property to PauliWord and PauliSentence then the first line in your example raises an error when passing paulis

Reply:
lol touché. I hacked together this example to try to get it going, but hadn't really thought on the correct order of operations. My concern is that `PauliWord` and `PauliSentence` were originally intended to be lightweight wrappers around dicts, and that's pretty much it. It's not the end of the world to add new properties and stuff to them (as we very much are doing now), but I'd want us to keep it as minimal as possible. Otherwise, PL descends into this everything-is-everything strategy and people make bad assumptions. eg. "this object has a `pauli_rep`, so it probably has a `matrix()` method too, because it sure looks like an Operator". We already have a similar issue with `MidMeasureMP` appearing in `tape.operations`, so PL thinks it satisfies the `Operator` API (it does not, but it's trying!).

That said, c'est la vie pythonique so I won't die on this particular hill if you'd prefer it this way

Reply:
> PS: did we call the matrix function to_mat on purpose, to avoid using matrix and mimic Operator?

I'd be curious to know the logic behind that naming as well 😅

Reply:
@Jaybsoni Can you remember? I remember thinking about naming it `matrix`, but it looks like we didn't have that discussion on the original PR.